shards/egui/src/visual/mod.rs

```diff
@@ -0,0 +1,2521 @@
+        identifier.accept_mut(self)
+      }),
+      Assignment::AssignSet(pipeline, identifier) => (pipeline.accept_mut(self), {
+        self.ui.label(emoji("➡"));
+        identifier.accept_mut(self)
+      }),
+      Assignment::AssignUpd(pipeline, identifier) => (pipeline.accept_mut(self), {
+        self.ui.label(emoji("⤴"));
+        identifier.accept_mut(self)
+      }),
+      Assignment::AssignPush(pipeline, identifier) => (pipeline.accept_mut(self), {
+        self.ui.label(emoji("⏩"));
+        identifier.accept_mut(self)
+      }),
+    };
+    if let Some(a) = r_a {
+      if let Some(b) = r_b {
+        combined_response = Some(a.union(b));
+      } else {
+        combined_response = Some(a);
+      }
+    }
+    combined_response.map(|x| {
+      if x.clicked() {
+        self.context.has_changed = true;
+      }
+      x
+    })
+  }
+
+  fn visit_pipeline(&mut self, pipeline: &mut Pipeline) -> Option<Response> {
+    let mut final_response: Option<Response> = None;
+    let mut i = 0;
+    while i < pipeline.blocks.len() {
+      let response = match self.visit_block(&mut pipeline.blocks[i]) {
+        (BlockAction::Remove, r) => {
+          self.context.has_changed = true;
+          pipeline.blocks.remove(i);
+          // if the blocks are empty, we should remove the pipeline
+          if pipeline.blocks.is_empty() {
+            let parent_sequence = self.context.seqs_stack.last().unwrap();
+            let parent_sequence = unsafe { &mut **parent_sequence };
+            // find pipeline index and remove it
+            let pipeline_index = parent_sequence
+              .statements
+              .iter()
+              .position(|x| match x {
+                Statement::Pipeline(x) => x as *const Pipeline == pipeline as *const Pipeline,
+                _ => false,
+              })
+              .unwrap();
+            parent_sequence.statements.remove(pipeline_index);
+          }
+          r
+        }
+        (BlockAction::Keep, r) => {
+          i += 1;
+          r
+        }
+        (BlockAction::Duplicate, r) => {
+          self.context.has_changed = true;
+          let mut block = pipeline.blocks[i].clone();
+          block.get_custom_state::<BlockState>().map(|x| {
+            x.id = Id::new(nanoid!(16));
+          });
+          pipeline.blocks.insert(i, block);
+          i += 2;
+          r
+        }
+        (BlockAction::Swap(block), r) => {
+          self.context.has_changed = true;
+          pipeline.blocks.get_mut(i).map(|x| {
+            let selected = x
+              .get_custom_state::<BlockState>()
+              .map(|x| x.selected)
+              .unwrap_or(false);
+            *x = block;
+            x.get_or_insert_custom_state(|| BlockState {
+              selected,
+              id: Id::new(nanoid!(16)),
+            })
+            .selected = selected;
+          });
+          i += 1;
+          r
+        }
+      };
+
+      // if let Some(response) = &response {
```

Comment:
Cleanup

Reply:
Was keeping those as memo on how to use it once I had to actually